api/client/trust.go

```diff
@@ -0,0 +1,437 @@
+	tokenHandler := auth.NewTokenHandler(authTransport, creds, repoInfo.CanonicalName, "push", "pull")
+	basicHandler := auth.NewBasicHandler(creds)
+	modifiers = append(modifiers, transport.RequestModifier(auth.NewAuthorizer(challengeManager, tokenHandler, basicHandler)))
+	tr := transport.NewTransport(base, modifiers...)
+
+	return client.NewNotaryRepository(cli.trustDirectory(), repoInfo.CanonicalName, server, tr, cli.getPassphraseRetriever())
+}
+
+func convertTarget(t client.Target) (target, error) {
+	h, ok := t.Hashes["sha256"]
+	if !ok {
+		return target{}, errors.New("no valid hash, expecting sha256")
+	}
+	return target{
+		reference: registry.ParseReference(t.Name),
+		digest:    digest.NewDigestFromHex("sha256", hex.EncodeToString(h)),
+		size:      t.Length,
+	}, nil
+}
+
+func (cli *DockerCli) getPassphraseRetriever() passphrase.Retriever {
+	baseRetriever := passphrase.PromptRetriever()
+	rootEnv := os.Getenv("DOCKER_TRUST_ROOT_PASSPHRASE")
+	targetEnv := os.Getenv("DOCKER_TRUST_TARGET_PASSPHRASE")
+	snapshotEnv := os.Getenv("DOCKER_TRUST_SNAPSHOT_PASSPHRASE")
+
+	return func(keyName string, alias string, createNew bool, numAttempts int) (string, bool, error) {
+		if alias == "root" && len(rootEnv) >= 8 {
+			return rootEnv, false, nil
+		}
+		if alias == "targets" && len(targetEnv) >= 8 {
+			return targetEnv, false, nil
+		}
+		if alias == "snapshot" && len(snapshotEnv) >= 8 {
+			return snapshotEnv, false, nil
+		}
+		return baseRetriever(keyName, alias, createNew, numAttempts)
+	}
+}
+
+func (cli *DockerCli) trustedReference(repo string, ref registry.Reference) (registry.Reference, error) {
+	repoInfo, err := registry.ParseRepositoryInfo(repo)
+	if err != nil {
+		return nil, err
+	}
+
+	// Resolve the Auth config relevant for this server
+	authConfig := registry.ResolveAuthConfig(cli.configFile, repoInfo.Index)
+
+	notaryRepo, err := cli.getNotaryRepository(repoInfo, authConfig)
+	if err != nil {
+		fmt.Fprintf(cli.out, "Error establishing connection to trust repository: %s\n", err)
+		return nil, err
+	}
+
+	t, err := notaryRepo.GetTargetByName(ref.String())
+	if err != nil {
+		return nil, err
+	}
+	r, err := convertTarget(*t)
+	if err != nil {
+		return nil, err
+
+	}
+
+	return registry.DigestReference(r.digest), nil
+}
+
+func (cli *DockerCli) tagTrusted(repoInfo *registry.RepositoryInfo, trustedRef, ref registry.Reference) error {
+	fullName := trustedRef.ImageName(repoInfo.LocalName)
+	fmt.Fprintf(cli.out, "Tagging %s as %s\n", fullName, ref.ImageName(repoInfo.LocalName))
+	tv := url.Values{}
+	tv.Set("repo", repoInfo.LocalName)
+	tv.Set("tag", ref.String())
+	tv.Set("force", "1")
+
+	if _, _, err := readBody(cli.call("POST", "/images/"+fullName+"/tag?"+tv.Encode(), nil, nil)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func notaryError(err error) error {
+	if _, ok := err.(*json.SyntaxError); ok {
+		logrus.Debugf("Notary syntax error: %s", err)
+		return errors.New("no trust data available for repository")
```

Comment:
agreed with @jfrazelle, we should add the original error to the returned one, or return the original one directly. It might also be useful to turn that `Debugf` into and `Errorf`